sympy/matrices/expressions/hadamard.py

```diff
@@ -119,11 +132,148 @@ def validate(*args):
-                                 do_one(*rules)))
+# TODO Implement algorithm for rewriting Hadamard product as diagonal matrix
+# if matmul identy matrix is multiplied.
+def canonicalize(x):
+    """Canonicalize the Hadamard product ``x`` with mathematical properties.
+
+    Examples
+    ========
+
+    >>> from sympy.matrices.expressions import MatrixSymbol, HadamardProduct
+    >>> from sympy.matrices.expressions import OneMatrix, ZeroMatrix
+    >>> from sympy.matrices.expressions.hadamard import canonicalize
+
+    >>> A = MatrixSymbol('A', 2, 2)
+    >>> B = MatrixSymbol('B', 2, 2)
+    >>> C = MatrixSymbol('C', 2, 2)
+
+    Hadamard product associativity:
+
+    >>> X = HadamardProduct(A, HadamardProduct(B, C))
+    >>> X
+    A.*(B.*C)
+    >>> canonicalize(X)
+    A.*B.*C
+
+    Hadamard product commutativity:
+
+    >>> X = HadamardProduct(A, B)
+    >>> Y = HadamardProduct(B, A)
+    >>> X
+    A.*B
+    >>> Y
+    B.*A
+    >>> canonicalize(X)
+    A.*B
+    >>> canonicalize(Y)
+    A.*B
+
+    Hadamard product identity:
+
+    >>> X = HadamardProduct(A, OneMatrix(2, 2))
+    >>> X
+    A.*OneMatrix(2, 2)
+    >>> canonicalize(X)
+    A
+
+    Absorbing element of Hadamard product:
+
+    >>> X = HadamardProduct(A, ZeroMatrix(2, 2))
+    >>> X
+    A.*0
+    >>> canonicalize(X)
+    0
+
+    Rewriting to Hadamard Power
+
+    >>> X = HadamardProduct(A, A, A)
+    >>> X
+    A.*A.*A
+    >>> canonicalize(X)
+    A.**3
+
+    Notes
+    =====
+
+    As Hadamard product is associative, all the nested products can be
+    flattened to the level 1.
+
+    As Hadamard product is commutative, all the arguments can be sorted to
+    the canonical form.
+
+    Matrix of only ones is an identity for Hadamard product,
+    so every ``OneMatrix`` can be removed.
+
+    Matrix of only zeros will make Hadamard product zero.
+
+    Duplicate elements can be collected and rewritten as HadamardPower
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Hadamard_product_(matrices)
+
+    .. [2] https://en.wikipedia.org/wiki/Multiplication
+    """
+    from sympy.core.compatibility import default_sort_key
+
+    # Associativity
+    rule = condition(
+            lambda x: isinstance(x, HadamardProduct),
+            flatten
+        )
+    fun = exhaust(rule)
+    x = fun(x)
+
+    # Identity
+    fun = condition(
+            lambda x: isinstance(x, HadamardProduct),
+            rm_id(lambda x: isinstance(x, OneMatrix))
+        )
+    x = fun(x)
+
+    # Absorbing by Zero Matrix
+    def absorb(x):
+        if any(isinstance(c, ZeroMatrix) for c in x.args):
+            return ZeroMatrix(*x.shape)
+        else:
+            return x
+    fun = condition(
+            lambda x: isinstance(x, HadamardProduct),
+            absorb
+        )
+    x = fun(x)
+
+    # Rewriting with HadamardPower
+    if isinstance(x, HadamardProduct):
+        tally = dict()
```

Comment:
Could use collections.Counter

Reply:
or `multiset` from `iterables`:
```
 >>> multiset('mississippi')
    {'i': 4, 'm': 1, 'p': 2, 's': 4}
```
so
```
tally = multiset(x.args)
```

Reply:
I've done some experiment in cpython 3.6
```python
from sympy.utilities.iterables import multiset
from collections import Counter

l = 'abc'+'c'*999999
```

`%timeit Counter(l)` `349 ms ± 53.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`
`%timeit multiset(l)` `578 ms ± 36.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`

I guess the one from the official library can use C function to speed up
https://github.com/python/cpython/blob/f2f55e7f03d332fd43bc665a86d585a79c3b3ed4/Modules/_collectionsmodule.c#L2237